Tasks/ServiceFabricUpdateManifestsV2/Update-DockerImageSettings.psm1

```diff
@@ -12,18 +12,44 @@ function Update-DockerImageSettings
         $appPackagePathSearchPattern = (Get-VstsInput -Name applicationPackagePath -Require)
         $appPackagePath = Find-VstsFiles -LegacyPattern $appPackagePathSearchPattern -IncludeDirectories
 
-        # Collect the image digest values for the images
+        $imageNamesFilePathSearchPattern = Get-VstsInput -Name imageNamesPath
+        $imageNamesPath = $null
+        if (-not [string]::IsNullOrEmpty($imageNamesFilePathSearchPattern))
+        {
+            $imageNamesPath = Find-VstsFiles -LegacyPattern $imageNamesFilePathSearchPattern
+        }
+
         $imageDigestValues = (Get-Content -LiteralPath $imageDigestsPath).Replace("`r`n", "`n").Split("`n")
+        $imageNames = $null
+        if (($imageNamesPath -ne $null))
+        {
+            $imageNames = (Get-Content -LiteralPath $imageNamesPath).Replace("`r`n", "`n").Split("`n")
+            if ($imageNames.Count -ne $imageDigestValues.Count)
+            {
+                throw (Get-VstsLocString -Key ImageDigestMismatch -ArgumentList @($imageNames.Count, $imageDigestValues.Count))
+            }
+        }
+
+        # Collect the image digest values for the images
         $imageNameToDigestMapping = @{}
-        foreach ($imageDigestValue in $imageDigestValues)
+        for ($ind = 0; $ind -lt $imageDigestValues.Count; $ind++)
         {
-            $slashIndex = $imageDigestValue.IndexOf("/")
-            $hashSeparatorIndex = $imageDigestValue.IndexOf("@")
-            if ($slashIndex -lt 0 -or $hashSeparatorIndex -lt 0)
+            $imageDigestValue = $imageDigestValues[$ind]
+            $imageName = $null
+            if ($imageNames -ne $null)
             {
-                throw (Get-VstsLocString -Key InvalidImageDigestValue -ArgumentList @($imageDigestValue, $imageDigestsPath))
+                $imageName = $imageNames[$ind]
+            }
+            else
+            {
+                $slashIndex = $imageDigestValue.IndexOf("/")
+                $hashSeparatorIndex = $imageDigestValue.IndexOf("@")
+                if ($slashIndex -lt 0 -or $hashSeparatorIndex -lt 0)
+                {
+                    throw (Get-VstsLocString -Key InvalidImageDigestValue -ArgumentList @($imageDigestValue, $imageDigestsPath))
+                }
+                $imageName = $imageDigestValue.Substring($slashIndex + 1, $hashSeparatorIndex - $slashIndex - 1)
             }
-            $imageName = $imageDigestValue.Substring($slashIndex + 1, $hashSeparatorIndex - $slashIndex - 1)
             $imageNameToDigestMapping[$imageName] = $imageDigestValue
```

Comment:
please add error message if during manifest update it fails to find this image name. User can then specify built image names file.

Reply:
Also, the BD template should be updated

Reply:
In adding the error checking I came up with what I think is a good compromise for implicitly handling the case where there is no ambiguity and giving an error when there are two services that collide.